Code/Entities/ClimbableOneWayPlatform.cs

```diff
@@ -230,35 +274,99 @@ public class ClimbableOneWayPlatform : Entity {
 
                 // inject ourselves to jump over the "Speed.Y < 0f" check, and put this back
                 cursor.EmitDelegate<Func<Player, bool>>(self => {
-                    ClimbableOneWayPlatform platform = collideFirstOutside(self, self.Position + new Vector2((int) self.Facing * 2, 0), self.Facing == Facings.Left);
+                    ClimbableOneWayPlatform platform = collideFirstOutside(self, self.Position + new Vector2((int)self.Facing * 2, 0), self.Facing == Facings.Left);
                     return platform != null && platform.climbJumpGrabCooldown <= 0f;
                 });
                 cursor.Emit(OpCodes.Brtrue, afterCheck);
                 cursor.Emit(OpCodes.Ldarg_0);
             }
         }
 
-        private static void modPlayerClimbJump(On.Celeste.Player.orig_ClimbJump orig, Player self) {
+        //grabs the speed and stamina if applicable at the earliest poossible time
+        private static void Customization_modPlayerClimbBegin(On.Celeste.Player.orig_ClimbBegin orig, Player self)
+        {
+            ClimbableOneWayPlatform platform = collideFirstOutside(self, self.Position + new Vector2((int)self.Facing * 2, 0), self.Facing == Facings.Left);
+            if (platform != null)
+            {
+                //store initial speed and stamina if needed   
+                if (platform.hasMomentumCarrying)
+                {
+                    platform.initialSpeed = self.Speed.X;
+                    platform.startTimer();
+                }
+                if (platform.initialStamina >= 0f)
+                {
+                    platform.stamStored = true;
+                    platform.initialStamina = self.Stamina;
+                }   
+            }
+            orig.Invoke(self);
+        }
+
+        private static int Customization_modPlayerClimbUpdate(On.Celeste.Player.orig_ClimbUpdate orig, Player self)
+        {
+
+            ClimbableOneWayPlatform platform = collideFirstOutside(self, self.Position + new Vector2((int)self.Facing * 2, 0), self.Facing == Facings.Left);
+            if (platform != null)
+            {
+                //checks for momentum jumps and same direction jumps. the latter cannot be activated if the former is active, as the purpose is nullified anyways
+                if((Input.Jump.Pressed || Input.Jump.BufferTime < 0.08f) && (int)Input.Aim.Value.X != 0) {
+                    if (platform.hasMomentumCarrying && platform.grabTimer < platform.decayTime) {
+                        self.Speed.X += platform.bonusSpeed(platform.initialSpeed, platform.grabTimer);
+                        platform.grabTimer = 0f;
+                    }
+                    else if(platform.sameDirBoost) {
+                        if (Math.Abs(self.Speed.X) < 240f && Input.Aim.Value.X / Math.Abs(Input.Aim.Value.X) == (int) self.Facing) {

```

Comment:
what happens to `Input.Aim.Value.X / Math.Abs(Input.Aim.Value.X)` if the direction is neutral (X = 0)?

0f/0f = NaN iirc so it probably ends up working but :a:

Reply:
this seems this could be replaced with `Math.Sign`, that can return -1, 0 or 1

Reply:
Using Math.Sign caused some problems involving weird not-a-number crashes without an error line, added a non-zero conditional to prevent weirdness. I didnt run into any crashes using neutral inputs with it, but this should work just in case.